### PR TITLE
refactor(database): move log logic into separate package

### PR DESCRIPTION
**Pull request by jbrockopp**
Based off of https://github.com/go-vela/server/pull/574, https://github.com/go-vela/server/pull/663, https://github.com/go-vela/server/pull/687, https://github.com/go-vela/server/pull/692 and https://github.com/go-vela/server/pull/721

This change continues the refactor efforts initially introduced in the above PRs.

This adds a new `log` package to the `github.com/go-vela/server/database` package.

This contains a `LogService` interface declaring all functions necessary for worker based interactions with the database:

https://github.com/go-vela/server/blob/6bf34484bbbf66c259bb6116132f5c72d47adeea/database/log/service.go#L11-L49

This package also contains the `engine` which implements the above service interface:

https://github.com/go-vela/server/blob/6bf34484bbbf66c259bb6116132f5c72d47adeea/database/log/log.go#L25-L39

This `engine` contains no raw SQL queries for integrating with the `logs` table.

Instead, we leverage our DB library's (https://gorm.io/) agnostic abstraction for integrating with that table.

### CODE REVIEW

File: database/log/list.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+	"github.com/go-vela/types/library"
+)
+
+// ListLogs gets a list of all logs from the database.
+func (e *engine) ListLogs() ([]*library.Log, error) {
+	e.logger.Trace("listing all logs from the database")
+
+	// variables to store query results and return value
+	count := int64(0)
+	h := new([]database.Log)
```
Comment (plyr4) [inline database/log/list.go:19]: does `h` signify something or is that copied from hooks? 
not a big deal

**Review by wass3r — APPROVED** [review]

**Review by plyr4 — APPROVED** [review]
LGTM thank you
I left some small nitpick comments that could be addressed separately, non-blocking

File: database/log/index.go [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+logs_build_id
+ON logs (build_id);
+`
+)
+
+// CreateLogIndexes creates the indexes for the logs table in the database.
+func (e *engine) CreateLogIndexes() error {
+	e.logger.Tracef("creating indexes for logs table in the database")
+
+	// create the hostname and address columns index for the logs table
```
[the remark this answers is not in the document]
Comment (jbrockopp) [reply database/log/index.go:22]: This has been addressed within https://github.com/go-vela/server/pull/780

File: database/log/get_step.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+	}
+
+	// decompress log data for the step
+	//
+	// https://pkg.go.dev/github.com/go-vela/types/database#Log.Decompress
+	err = l.Decompress()
+	if err != nil {
+		// ensures that the change is backwards compatible
+		// by logging the error instead of returning it
+		// which allowing us to fetch uncompressed logs
```
[the remark this answers is not in the document]
Comment (jbrockopp) [reply database/log/get_step.go:38]: This has been addressed within https://github.com/go-vela/server/pull/780

File: database/log/get_service.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+	}
+
+	// decompress log data for the service
+	//
+	// https://pkg.go.dev/github.com/go-vela/types/database#Log.Decompress
+	err = l.Decompress()
+	if err != nil {
+		// ensures that the change is backwards compatible
+		// by logging the error instead of returning it
+		// which allowing us to fetch uncompressed logs
```
[the remark this answers is not in the document]
Comment (jbrockopp) [reply database/log/get_service.go:38]: This has been addressed within https://github.com/go-vela/server/pull/780

File: database/log/list.go (hunk 1 above)
Comment (jbrockopp) [reply database/log/list.go:19]: This has been addressed within https://github.com/go-vela/server/pull/780

**Pull request merged**